sdks/java/extensions/schemaio-expansion-service/src/main/java/org/apache/beam/sdk/extensions/schemaio/expansion/ExternalSchemaIOTransformRegistrar.java

```diff
@@ -0,0 +1,142 @@
+        builder.put(
+            "beam:external:java:" + schemaIOProvider.identifier() + ":write:v1",
+            new WriterBuilder(schemaIOProvider));
+      }
+    } catch (Exception e) {
+      throw new RuntimeException(e.getMessage());
+    }
+    return builder.build();
+  }
+
+  public static class Configuration {
+    String location = "";
+    byte[] config = new byte[0];
+    @Nullable byte[] dataSchema = null;
+
+    public void setLocation(String location) {
+      this.location = location;
+    }
+
+    public void setConfig(byte[] config) {
+      this.config = config;
+    }
+
+    public void setDataSchema(byte[] dataSchema) {
+      this.dataSchema = dataSchema;
+    }
+  }
+
+  @Nullable
+  private static Schema translateSchema(@Nullable byte[] schemaBytes) throws Exception {
+    if (schemaBytes == null) {
+      return null;
+    }
+    SchemaApi.Schema protoSchema = SchemaApi.Schema.parseFrom(schemaBytes);
+    return SchemaTranslation.schemaFromProto(protoSchema);
+  }
+
+  private static Row translateRow(byte[] rowBytes, Schema configSchema) throws Exception {
+    RowCoder rowCoder = RowCoder.of(configSchema);
+    InputStream stream = new ByteArrayInputStream(rowBytes);
+    return rowCoder.decode(stream);
+  }
+
+  private static class ReaderBuilder
+      implements ExternalTransformBuilder<Configuration, PBegin, PCollection<Row>> {
+    SchemaIOProvider schemaIOProvider;
+
+    ReaderBuilder(SchemaIOProvider schemaIOProvider) {
+      this.schemaIOProvider = schemaIOProvider;
+    }
+
+    @Override
+    public PTransform<PBegin, PCollection<Row>> buildExternal(Configuration configuration) {
+      try {
+        return schemaIOProvider
+            .from(
+                configuration.location,
+                translateRow(configuration.config, schemaIOProvider.configurationSchema()),
+                translateSchema(configuration.dataSchema))
+            .buildReader();
+      } catch (Exception e) {
+        throw new RuntimeException("Could not convert configuration proto to row or schema.");
+      }
+    }
+  }
+
+  private static class WriterBuilder
+      implements ExternalTransformBuilder<Configuration, PCollection<Row>, PDone> {
+    SchemaIOProvider schemaIOProvider;
+
+    WriterBuilder(SchemaIOProvider schemaIOProvider) {
+      this.schemaIOProvider = schemaIOProvider;
+    }
+
+    @Override
+    public PTransform<PCollection<Row>, PDone> buildExternal(Configuration configuration) {
+      try {
+        return (PTransform<PCollection<Row>, PDone>)
+            schemaIOProvider
+                .from(
+                    configuration.location,
+                    translateRow(configuration.config, schemaIOProvider.configurationSchema()),
+                    translateSchema(configuration.dataSchema))
+                .buildWriter();
+      } catch (Exception e) {
+        throw new RuntimeException("Could not convert configuration proto to row or schema.");
```

Comment:
This isn't ideal since it will swallow up any other exception that occurs inside the try block, and when it does catch something from translateRow or translateSchema it's not clear where the problem is. Could you instead add a try-catch in translateRow and translateSchema that catches the specific exception(s) that might be thrown, and re-throws them with a RuntimeException?

When you re-throw you should always reference the original exception as well, like `throw new RuntimeException("Some additional context", e)`

Reply:
Done